Install more in bootstrap

mikemcquaid opened the pull request:
Currently there's assumptions elsewhere that `rbenv` and `bundle` are used so `bootstrap` should probably also check to see if they are installed. Also, if you're on OS X use `brew bundle` to install Homebrew dependencies from a `Brewfile`.

CC @maddox for review. Feel free to reject this if you feel it's overcomplicating things; no hard feelings if so.

technicalpickles commented on script/bootstrap:
```diff
@@ -7,7 +7,22 @@ set -e
   bundle install --path vendor/gems --quiet --without production
 }
+
+if [ "$(uname -s)" = "Darwin" ] && brew --prefix 2>&1 > /dev/null; then
```
Should this check for the presence of a `Brewfile` too?

mikemcquaid replied:
I'm assuming not given it doesn't currently check for a `Gemfile.`

technicalpickles commented on script/bootstrap:
```diff
@@ -7,7 +7,22 @@ set -e
+  echo "==> Installing Homebrew dependencies…"
+  brew update
+  brew tap homebrew/bundle 2>/dev/null
+  brew bundle check 2>&1 > /dev/null || {
```
This exits with 1 if there isn't a Brewfile, so and `brew bundle` fails when it's not present:

```
$ brew bundle
/opt/boxen/homebrew/Library/Taps/homebrew/homebrew-bundle/lib/bundle/utils.rb:35:in `rescue in brewfile': No Brewfile found (RuntimeError)
        from /opt/boxen/homebrew/Library/Taps/homebrew/homebrew-bundle/lib/bundle/utils.rb:28:in `brewfile'
        from /opt/boxen/homebrew/Library/Taps/homebrew/homebrew-bundle/lib/bundle/commands/install.rb:4:in `run'
        from /opt/boxen/homebrew/Library/Taps/homebrew/homebrew-bundle/cmd/brew-bundle:66:in `<main>'

```

mikemcquaid replied:
Sure, does that not apply to `bundle check` and `bundle install` above, though?

technicalpickles replied:
It does. But ust because there is a not-great thing in this file already additions shouldn't try to be better when adding new.

maddox replied:
> It does. But ust because there is a not-great thing in this file already additions shouldn't try to be better when adding new.

I'm hip to that.

mikemcquaid replied:
So, my problem with that is that this file/repository assumes a Gemfile exists in the root directory even although one is included in the repo. If you didn't have a `Gemfile` because this was e.g. a NodeJS project then you wouldn't add a `Gemfile` check, you'd just not include this code. This feels like the same thing to me with the `Brewfile`: if you weren't using one for this repository you'd just not include this code rather than guard it to be a no-op if it doesn't exist.

I think this differs from the OS check because you may well have a `Brewfile` present for Linux machines but don't want it run (and there's no obvious Linux alternative yet as far as I can tell).

mikemcquaid replied:
@maddox @technicalpickles Any more thoughts on this? Happy to either add a check for the `Gemfile` or leave it without a `Brewfile` check but feels odd to have them inconsistent.

technicalpickles replied:
How about leaving it as is here (without the checks) and follow up with a PR to add checks more consistently?

maddox replied:
It's also that it will force install bundler to handle gem deps, but won't install homebrew to handle brew deps.

I feel like there's mixed messages here on how much of the environment is handled.

mikemcquaid replied:
@maddox I'm thinking it installs bundler because that's a dependency specific to a Ruby version (as this repo may use a Ruby version that no others do) whereas Homebrew is a system-level dependency that may have a variety of locations.

maddox commented on script/bootstrap:
```diff
@@ -7,7 +7,22 @@ set -e
 
 cd "$(dirname "$0")/.."
 
+echo "==> Installing Ruby…"
```
I'm weary of this `echo`, because...while the other steps echo what they're doing, they actually ARE doing it. When they complete fast, it tells the user that (and it's because) "oh cool, it did it and its done fast" or "it didn't have to do anything." 

Seeing...
    Installing Ruby...
...every time isn't just scary, I think it's confusing.

mikemcquaid replied:
Ok, cool.

mikemcquaid commented:
Have addressed all comments. From my commit message (to save you a click):

- Guard actions based on presence of `Brewfile`, `.ruby-version`, `Gemfile`.
- Homebrew's dependencies should be installed first as they may be installing e.g. `rbenv`.
- `rbenv install` is only needed if the desired Ruby version can't be found.
- `rbenv rehash` may be needed after `bundler` is installed.

maddox commented:
@mikemcquaid This is looking great. `script/bootstrap` is getting 'thicker' but at the expense of NO BOXEN :thumbsup:

I especially like how all these sections read now. The guard, report, do stuff patterns are easier on the eyes now and way easier to grok.

maddox commented:
Maybe add a little to [this section](https://github.com/github/scripts-to-rule-them-all#scriptbootstrap) of the README?

> brew packages, ruby versions

Otherwise, I'm :thumbsup: to :ship:.

mikemcquaid commented:
README modified, :ship:d!
